drivers/serial/uart_mcux_flexcomm.c

```diff
@@ -342,6 +379,636 @@ static int mcux_flexcomm_uart_config_get(const struct device *dev,
 }
 #endif /* CONFIG_UART_USE_RUNTIME_CONFIGURE */
 
+#ifdef CONFIG_UART_ASYNC_API
+/* This function is called by this driver to notify user callback of events */
+static void async_user_callback(const struct device *dev,
+					      struct uart_event *evt)
+{
+	const struct mcux_flexcomm_data *data = dev->data;
+
+	if (data->async_callback) {
+		data->async_callback(dev, evt, data->async_cb_data);
+	}
+}
+
+static int mcux_flexcomm_uart_callback_set(const struct device *dev,
+					   uart_callback_t callback,
+					   void *user_data)
+{
+	struct mcux_flexcomm_data *data = dev->data;
+
+	data->async_callback = callback;
+	data->async_cb_data = user_data;
+
+	return 0;
+}
+
+static int mcux_flexcomm_uart_tx(const struct device *dev, const uint8_t *buf,
+				 size_t len, int32_t timeout)
+{
+	const struct mcux_flexcomm_config *config = dev->config;
+	struct mcux_flexcomm_data *data = dev->data;
+	int ret = 0;
+
+	if (config->tx_dma.dev == NULL) {
+		return -ENODEV;
+	}
+
+	unsigned int key = irq_lock();
+
+	/* Getting DMA status to tell if channel is busy or not set up */
+	struct dma_status status;
+
+	ret = dma_get_status(config->tx_dma.dev, config->tx_dma.channel, &status);
+
+	if (ret < 0) {
+#ifdef CONFIG_DMA_MCUX_LPC
```

Comment:
The custom behavior here based on which DMA driver is being used concerns me. I assume that the first time we attempt to transmit using this channel it would be unconfigured, and  we would get the `-EACCES` return code.


Rather than using this driver specific behavior, could you instead track if the TX is busy using something like the state of `data->tx_data.xfer_len`, or another variable? It seems based on the implementation here that `data->tx_data.xfer_len` would only be 0 when a transfer is not active (you of course would need to initialized `data->tx_data.xfer_len = 0` in the driver init for this to work)

Reply:
Well I think maybe not because this channel can also be used for DMIC requests and memory to memory, and I don't want to make any assumptions about how a user is or isn't writing their application regarding reconfiguring and re-using DMA channels at runtime. And you can't just check if tx_data.xfer_len is 0 && ret = 0, because we need to differentiate the -eacces code from the -ebusy

Also, this is more a problem with that DMA driver, maybe it shouldn't be returning -EACCES, maybe it should just say "not busy" if a channel is not set up to do anything. The EACCES code is not part of the dma api, its a custom error code, it was added because of something like the dma driver interpreting the channel argument as a "virtual channel" in the dma driver and lots of k_mallocs and  garbage values, I don't remember those details too well now, it's a bigger problem with that DMA driver imo that it shouldnt be using that virtual channel scheme.

Reply:
I think it would be better to fix the bug in the DMA driver